src/Validate.tsx

```diff
@@ -0,0 +1,88 @@
+import React, {KeyboardEvent, RefObject} from "react";
+import {Contradictions, Groupings, SquareAddress, SquareValue, Values} from "./App";
+
+// tslint:disable-next-line:no-var-requires
+const _ = require("lodash");
+
+export type Contradicts = boolean;
+
+ /*
+ * Contradictions should only contain contradicting mappings
+ */
+
+const ROWS =   [[0 , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8],
+                [9 , 10, 11, 12, 13, 14, 15, 16, 17],
+                [18, 19, 20, 21, 22, 23, 24, 25, 26],
+                [27, 28, 29, 30, 31, 32, 33, 34, 35],
+                [36, 37, 38, 39, 40, 41, 42, 43, 44],
+                [45, 46, 47, 48, 49, 50, 51, 52, 53],
+                [54, 55, 56, 57, 58, 59, 60, 61, 62],
+                [63, 64, 65, 66, 67, 68, 69, 70, 71],
+                [72, 73, 74, 75, 76, 77, 78, 79, 80]];
+
+const COLUMNS = [[0, 9 , 18, 27, 36, 45, 54, 63, 72],
+                 [1, 10, 19, 28, 37, 46, 55, 64, 73],
+                 [2, 11, 20, 29, 38, 47, 56, 65, 74],
+                 [3, 12, 21, 30, 39, 48, 57, 66, 75],
+                 [4, 13, 22, 31, 40, 49, 58, 67, 76],
+                 [5, 14, 23, 32, 41, 50, 59, 68, 77],
+                 [6, 15, 24, 33, 42, 51, 60, 69, 78],
+                 [7, 16, 25, 34, 43, 52, 61, 70, 79],
+                 [8, 17, 26, 35, 44, 53, 62, 71, 80]];
```

Comment:
I would eliminate the concept of `ROWS` and `COLUMNS` and instead have a top-level const called groupings.

```js
const groupings: Groupings = [
  // Row Groupings
  [0 , 1 , 2 , 3 , 4 , 5 , 6 , 7 , 8],
  // ...
  // Column Groupings
  [1, 10, 19, 28, 37, 46, 55, 64, 73],
  // ...
  // Square Groupings
  [0, 1, 2, 9, 10, 11, 18, 19, 20],
  // ...
]
```

Reply:
I organized it like this because in the future the `ROWS` and `COLUMNS` won't change but the groupings might. So that's why groupings are passed in rather than also maintained as a constant.

Reply:
You're over-optimizing for a future use case. You shouldn't preemptively generify your code until you actually have your second use case.

After reading your tests, this should be:

```js
export const normalSudokuGroupings: Groupings = ...
```

This way, you can use normalSudokuGroupings in your test.

Reply:
You're over-optimizing for a future use case. You shouldn't preemptively generify your code until you actually have your second use case.

After reading your tests, this should be:

```js
export const normalSudokuGroupings: Groupings = ...
```

This way, you can use normalSudokuGroupings in your test.